inference/conflict.go

```diff
@@ -19,148 +19,177 @@ import (
+	return fmt.Sprintf(" Nonnil `%s` expected at \"%s\", but produced as nilable at \"%s\". Observed nil flow from "+
+		"source to dereference: %s", c.expr, consumerPos.String(), producerPos.String(), c.nilFlow.String())
 }
 
-// getPosition returns the source code position of the conflict.
-func (t *singleAssertionConflict) getPosition() token.Pos {
-	return t.trigger.Pos
+type conflictList struct {
+	conflicts []conflict
 }
 
-// getProducerMessage returns the error message to be reported for the producer part of the conflict.
-func (t *singleAssertionConflict) getProducerMessage() string {
-	return fmt.Sprintf(" Value %s (definitely nilable)", t.trigger.ProducerRepr)
-}
+func (l *conflictList) addSingleAssertionConflict(pass *analysis.Pass, trigger annotation.FullTrigger) {
+	t := fullTriggerAsPrimitive(pass, trigger)
+	c := conflict{
+		position: t.Pos,
+		expr:     util.ExprToString(trigger.Consumer.Expr, pass),
+		nilFlow:  nilFlow{},
+	}
 
-// getConsumerMessage returns the error message to be reported for the consumer part of the conflict.
-func (t *singleAssertionConflict) getConsumerMessage() string {
-	return fmt.Sprintf(" and is %s (must be nonnil)", t.trigger.ConsumerRepr)
-}
+	c.nilFlow.addNilPathNode(t.ProducerRepr)
+	c.nilFlow.addNonNilPathNode(t.ConsumerRepr)
 
-// key returns the key of the conflict group that this conflict belongs to.
-func (t *singleAssertionConflict) key() groupKey {
-	return t.originalTrigger.Producer
+	l.conflicts = append(l.conflicts, c)
 }
 
-// An overconstrainedConflict represents a local annotation site that was constrained by two different
-// chains of assertions to be both nilable (true) and nonnil (false). It encodes the overconstrained
-// site, and the reason that the site had to be true and had to be false, as ExplainedBools.
-type overconstrainedConflict struct {
-	site             primitiveSite
-	trueExplanation  ExplainedBool
-	falseExplanation ExplainedBool
-}
+func (l *conflictList) addOverconstraintConflict(nilExplanation ExplainedBool, nonnilExplanation ExplainedBool, pass *analysis.Pass) {
+	c := conflict{}
 
-// newOverconstrainedConflict  constructs and returns an overconstrainedConflict.
-func newOverconstrainedConflict(site primitiveSite, trueExplanation, falseExplanation ExplainedBool) *overconstrainedConflict {
-	return &overconstrainedConflict{
-		site:             site,
-		trueExplanation:  trueExplanation,
-		falseExplanation: falseExplanation,
+	// Build nil path by traversing the inference graph from `nilExplanation` part of the overconstraint failure.
+	// (Note that this traversal gives us a backward path from point of conflict to the source of nilability. Hence, we
+	// must take this into consideration while printing the flow.)
+	var queue []ExplainedBool
+	queue = append(queue, nilExplanation)
+
+	for len(queue) > 0 {
+		e := queue[0]
+		queue = queue[1:]
+
+		t := e.getPrimitiveFullTrigger()
+		if t.ConsumerRepr != nil && t.ProducerRepr != nil {
```

Comment:
Can I get a one sentence reminder in comments here about what these two cases for `t` represent? :)

Reply:
Yes, added a comment :)